src/pudl_archiver/archivers/eia/eiarecs.py

```diff
@@ -0,0 +1,115 @@
+"""Archive EIA Residential Energy Consumption Survey (RECS)."""
+
+import logging
+import re
+
+from pudl_archiver.archivers.classes import (
+    AbstractDatasetArchiver,
+    ArchiveAwaitable,
+    ResourceInfo,
+)
+from pudl_archiver.frictionless import ZipLayout
+
+LINK_PATTERNS = [
+    # housing characteristics
+    {
+        "base_url": "https://www.eia.gov/consumption/residential/data",
+        "php_extension": "index.php?view=characteristics",
+        "prefix": "hc",
+        "pattern": re.compile(r"HC (\d{1,2})\.(\d{1,2})\.xlsx"),
+    },
+    # consumption & expenditures
+    {
+        "base_url": "https://www.eia.gov/consumption/residential/data",
+        "php_extension": "index.php?view=consumption",
+        "prefix": "ce",
+        "pattern": re.compile(r"ce(\d)\.(\d{1,2})([a-z]?)\.xlsx"),
+    },
+    # state data (housing characteristics)
+    {
+        "base_url": "https://www.eia.gov/consumption/residential/data",
+        "php_extension": "index.php?view=state",
+        "prefix": "state",
+        "pattern": re.compile(r"State (.*)\.xlsx"),
+    },
+    # state data (consumption & expenditures)
+    {
+        "base_url": "https://www.eia.gov/consumption/residential/data",
+        "php_extension": "index.php?view=state",
+        "prefix": "state-ce",
+        "pattern": re.compile(r"ce(\d)\.(\d{1,2})\.(.*)\.xlsx"),
+    },
+    # microdata
+    # adding this in will require major changes+cleanup to the code below
+    # {
+    #    "base_url": "https://www.eia.gov/consumption/residential/data",
+    #    "php_extension": "index.php?view=microdata",
+    #    "prefix": "udata",
+    #    "pattern": re.compile(r"(recs.*\d{4}.*public.*)\.(?:zip|csv|xlsx)", re.IGNORECASE),
+    # }
+]
+logger = logging.getLogger(f"catalystcoop.{__name__}")
+
+
+class EiaRECSArchiver(AbstractDatasetArchiver):
+    """EIA RECS archiver."""
+
+    name = "eiarecs"
+
+    async def get_resources(self) -> ArchiveAwaitable:
+        """Download EIA-RECS resources."""
+        for year in [2020]:
+            yield self.get_year_resources(year)
+
+    async def get_year_resources(self, year: int) -> list[ResourceInfo]:
+        """Download all excel tables for a year."""
+        # Loop through all download links for tables
+        tables = []
+        zip_path = self.download_directory / f"eia-recs-{year}.zip"
+        data_paths_in_archive = set()
+        # Loop through different categories of data (all .xlsx)
+        for pattern_dict in LINK_PATTERNS:
```

Comment:
We call `get_hyperlinks` multiple times: once for each tab of data.

Reply:
What a fun webpage!